packages/blocks/src/page-block/edgeless/toolbar/edgeless-toolbar.ts

```diff
@@ -115,14 +116,41 @@ export class EdgelessToolbar extends LitElement {
     this.edgeless?.slots.mouseModeUpdated.emit(mouseMode);
   }
 
-  private _setZoom(zoom: number) {
-    const { viewport } = this.edgeless.surface;
-    viewport.setZoom(zoom);
+  private _setCenter(x: number, y: number) {
+    this.edgeless.surface.viewport.setCenter(x, y);
+    this.edgeless.slots.viewportUpdated.emit();
+  }
+
+  private _setZoom(zoom: number, focusPoint?: IPoint) {
+    this.edgeless.surface.viewport.setZoom(zoom, focusPoint);
     this.edgeless.slots.viewportUpdated.emit();
   }
 
   private _setZoomByStep(step: number) {
-    this._setZoom(clamp(this.zoom + step, ZOOM_MIN, ZOOM_MAX));
+    this._smoothZoom(clamp(this.zoom + step, ZOOM_MIN, ZOOM_MAX));
+  }
+
+  private _smoothZoom(zoom: number, focusPoint?: Point) {
+    const delta = zoom - this.zoom;
+
+    const innerSmoothZoom = () => {
+      requestAnimationFrame(() => {
+        const sign = delta > 0 ? 1 : -1;
+        const total = 10;
+        const step = delta / total;
```

Comment:
It seems to be able to move outside

Reply:
there is a _cutoff method to ensure it will not exceed.